File: plugin/logger.py
from datetime import datetime
from pathlib import Path
import traceback


def log(plugin_dir: Path, message: str):
    try:
        data_dir = plugin_dir / "Data"
        data_dir.mkdir(exist_ok=True)
        line = f"{datetime.now().strftime('%Y-%m-%d %H:%M:%S.%f')[:-3]} {message}\n"
        with (data_dir / "paste_tool.log").open("a", encoding="utf-8") as handle:
            handle.write(line)
    except Exception:
        pass


def log_exception(plugin_dir: Path, message: str):
    try:
        log(plugin_dir, f"{message}\n{traceback.format_exc()}")
    except Exception:
        pass

File: Logger.cs
using System;
using System.IO;

namespace Flow.Launcher.Plugin.PasteTool;

internal static class Logger
{
    private static string? _logFile;
    private static readonly object _lock = new();

    public static void Init(string pluginDir)
    {
        var dataDir = Path.Combine(pluginDir, "Data");
        Directory.CreateDirectory(dataDir);
        _logFile = Path.Combine(dataDir, "paste_tool.log");
    }

    public static void Log(string message)
    {
        if (_logFile == null) return;
        try
        {
            lock (_lock)
            {
                File.AppendAllText(_logFile, $"{DateTime.Now:yyyy-MM-dd HH:mm:ss.fff} {message}{Environment.NewLine}");
            }
        }
        catch { }
    }

    public static void LogException(string message, Exception ex)
    {
        Log($"{message}\n{ex}");
    }
}
